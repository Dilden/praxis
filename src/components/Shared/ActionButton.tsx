import {
  ButtonProps,
  createStyles,
  makeStyles,
  Button,
} from "@material-ui/core";

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      textTransform: "none",
    },
  })
);

const ActionButton = (props: ButtonProps) => {
  const classes = useStyles();

  return (
    <Button color="primary" classes={classes} {...props}>
      {props.children}
    </Button>
  );
};

export default ActionButton;
