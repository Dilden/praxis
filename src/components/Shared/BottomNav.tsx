import { useEffect, useState } from "react";
import Link from "next/link";
import Router, { useRouter } from "next/router";
import { BottomNavigation, BottomNavigationAction } from "@material-ui/core";
import { Person, Home, Group, Menu } from "@material-ui/icons";

import { NavigationPaths, ResourcePaths } from "../../constants/common";
import { navOpenVar } from "../../apollo/client/localState";
import { useCurrentUser } from "../../hooks";
import Messages from "../../utils/messages";
import { useReactiveVar } from "@apollo/client";

interface NavLinkProps {
  href: string;
  text?: string;
  icon?: React.ReactChild;
}

const NavLink = ({ href, icon, text }: NavLinkProps) => {
  return (
    <Link href={href}>
      <a style={{ color: "inherit" }}>{icon ? icon : text}</a>
    </Link>
  );
};

const BottomNav = () => {
  const currentUser = useCurrentUser();
  const [value, setValue] = useState<number>(0);
  const navDrawerOpen = useReactiveVar(navOpenVar);
  const userProfilePath = `${ResourcePaths.User}${currentUser?.name}`;
  const { asPath: currentPath } = useRouter();

  useEffect(() => {
    if (!navDrawerOpen)
      switch (currentPath) {
        case NavigationPaths.Home:
          setValue(0);
          break;
        case getMatching(ResourcePaths.Group):
        case NavigationPaths.Groups:
          setValue(1);
          break;
        case userProfilePath:
          setValue(2);
          break;
        default:
          setValue(3);
      }
  }, [currentPath, navDrawerOpen]);

  const getMatching = (path: string): string => {
    const match = currentPath.match(path);
    if (match) return currentPath;
    return "";
  };

  return (
    <BottomNavigation
      value={value}
      onChange={(_event, newValue) => {
        setValue(newValue);
      }}
      showLabels
    >
      <BottomNavigationAction
        onClick={() => Router.push(NavigationPaths.Home)}
        icon={<NavLink href={NavigationPaths.Home} icon={<Home />} />}
        label={Messages.navigation.home()}
      />

      <BottomNavigationAction
        onClick={() => Router.push(NavigationPaths.Groups)}
        icon={<NavLink href={NavigationPaths.Groups} icon={<Group />} />}
        label={Messages.navigation.groups()}
      />

      <BottomNavigationAction
        disabled={!currentUser}
        onClick={() => Router.push(userProfilePath)}
        icon={<NavLink href={userProfilePath} icon={<Person />} />}
        label={Messages.navigation.profile()}
      />

      <BottomNavigationAction
        onClick={() => navOpenVar(true)}
        label={Messages.navigation.menu()}
        icon={<Menu />}
      />
    </BottomNavigation>
  );
};

export default BottomNav;
