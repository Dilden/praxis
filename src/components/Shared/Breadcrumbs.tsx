import { useReactiveVar } from "@apollo/client";
import { Breadcrumbs, makeStyles, Typography } from "@material-ui/core";
import Link from "next/link";

import { breadcrumbsVar } from "../../apollo/client/localState";

const gray = "rgb(190, 190, 190)";

const useStyles = makeStyles({
  root: {
    marginBottom: 18,
  },
  separator: {
    color: gray,
  },
});

const CommonBreadcrumbs = () => {
  const breadcrumbs = useReactiveVar(breadcrumbsVar);
  const classes = useStyles();

  return (
    <Breadcrumbs classes={{ root: classes.root, separator: classes.separator }}>
      {breadcrumbs.map(({ label, href }) => {
        if (href)
          return (
            <Link href={href}>
              <a>
                <Typography style={{ color: gray }} key={label}>
                  {label}
                </Typography>
              </a>
            </Link>
          );
        return (
          <Typography style={{ color: "white" }} key={label}>
            {label}
          </Typography>
        );
      })}
    </Breadcrumbs>
  );
};

export default CommonBreadcrumbs;
