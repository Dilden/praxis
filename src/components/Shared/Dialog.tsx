import { forwardRef } from "react";
import {
  Button,
  makeStyles,
  Slide,
  Dialog,
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  Container,
} from "@material-ui/core";
import { Close } from "@material-ui/icons";
import { TransitionProps } from "@material-ui/core/transitions";
import { CircularProgress } from "@material-ui/core";

const useStyles = makeStyles({
  titleWithCaption: {
    flex: 1,
  },
  appBar: {
    position: "relative",
    backgroundColor: "rgb(30, 30, 30)",
    marginBottom: 40,
  },
});

const DialogTransition = forwardRef(function Transition(
  props: TransitionProps & { children?: React.ReactElement },
  ref: React.Ref<unknown>
) {
  return <Slide direction="up" ref={ref} {...props} />;
});

interface Props {
  title: string;
  subtext?: string;
  actionLabel?: string;
  closingAction?: (_: any) => any;
  open: boolean;
  setOpen: (open: boolean) => void;
  children: React.ReactChild;
  loading?: boolean;
}

const CommonDialog = ({
  title,
  subtext,
  open,
  setOpen,
  actionLabel,
  closingAction,
  children,
  loading,
}: Props) => {
  const classes = useStyles();

  const handleCloseDialog = () => {
    setOpen(false);
  };

  return (
    <Dialog
      open={open}
      TransitionComponent={DialogTransition}
      onClose={handleCloseDialog}
      PaperProps={{
        style: {
          backgroundColor: "rgb(50, 50, 50)",
        },
      }}
      fullScreen
    >
      <AppBar className={classes.appBar}>
        <Toolbar>
          <IconButton color="inherit" onClick={handleCloseDialog}>
            <Close />
          </IconButton>
          <div className={classes.titleWithCaption}>
            <Typography
              variant="h6"
              style={subtext ? { marginTop: 5, marginBottom: -10 } : {}}
            >
              {title}
            </Typography>
            <Typography
              variant="caption"
              style={{ color: "rgb(150, 150, 150)" }}
            >
              {subtext}
            </Typography>
          </div>
          {actionLabel && closingAction && !loading && (
            <Button color="inherit" onClick={closingAction}>
              {actionLabel}
            </Button>
          )}
          {loading && <CircularProgress style={{ color: "white" }} />}
        </Toolbar>
      </AppBar>
      <Container maxWidth="md">{children}</Container>
    </Dialog>
  );
};

export default CommonDialog;
