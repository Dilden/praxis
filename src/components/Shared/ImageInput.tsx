import { ChangeEvent, useRef } from "react";
import { Image } from "@material-ui/icons";

import styles from "../../styles/Shared/Shared.module.scss";

interface Props {
  setImage?: (image: File) => void;
  setImages?: (images: File[]) => void;
  multiple?: boolean;
  refreshKey: string;
}

const ImageInput = ({ setImage, setImages, multiple, refreshKey }: Props) => {
  const imageInput = useRef<HTMLInputElement>(null);

  const setImageState = (files: File[]) => {
    if (multiple && setImages) setImages(files);
    else if (setImage) setImage(files[0]);
  };

  return (
    <>
      <input
        multiple={multiple}
        type="file"
        accept="image/*"
        ref={imageInput}
        key={refreshKey}
        onChange={(e: ChangeEvent<HTMLInputElement>) =>
          e.target.files && setImageState([...e.target.files])
        }
        className={styles.imageInput}
      />
      <Image
        className={styles.imageInputIcon}
        onClick={() => imageInput.current?.click()}
        fontSize="large"
      />
    </>
  );
};

export default ImageInput;
