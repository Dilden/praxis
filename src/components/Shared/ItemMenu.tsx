import { Edit, Delete, MoreVert } from "@material-ui/icons";
import { Menu, MenuItem, IconButton } from "@material-ui/core";
import Link from "next/link";

interface Props {
  name?: string;
  itemId: string;
  itemType: string;
  anchorEl: null | HTMLElement;
  setAnchorEl: (el: null | HTMLElement) => void;
  deleteItem: (id: string) => void;
  ownItem: () => boolean;
  children?: React.ReactNode;
}

const ItemMenu = ({
  name,
  itemId,
  itemType,
  anchorEl,
  setAnchorEl,
  deleteItem,
  ownItem,
  children,
}: Props) => {
  const handleMenuButtonClick = (
    event: React.MouseEvent<HTMLButtonElement>
  ) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  if (ownItem())
    return (
      <>
        <IconButton onClick={handleMenuButtonClick}>
          <MoreVert style={{ color: "white" }} />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          keepMounted
          open={Boolean(anchorEl)}
          onClose={handleClose}
          PaperProps={{
            style: {
              backgroundColor: "rgb(65, 65, 65)",
            },
          }}
          MenuListProps={{
            style: {
              color: "white",
            },
          }}
        >
          <MenuItem>
            <Link href={`/${itemType}s/${name ? name : itemId}/edit`}>
              <a style={{ width: "100%" }}>
                <Edit
                  fontSize="small"
                  style={{
                    marginRight: "5",
                    transform: "rotateY(180deg)",
                  }}
                />
                Edit
              </a>
            </Link>
          </MenuItem>
          <MenuItem
            onClick={() =>
              window.confirm(
                `Are you sure you want to delete this ${itemType}?`
              ) && deleteItem(itemId)
            }
          >
            <Delete fontSize="small" style={{ marginRight: "5" }} />
            Delete
          </MenuItem>

          {children}
        </Menu>
      </>
    );

  return <></>;
};

export default ItemMenu;
