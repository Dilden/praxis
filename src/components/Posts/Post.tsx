import { useEffect, useState } from "react";
import { useRouter } from "next/router";
import { useQuery, useLazyQuery } from "@apollo/client";
import Link from "next/link";
import {
  Card,
  CardContent,
  Typography,
  makeStyles,
  CardHeader,
  CardMedia,
  CardActionArea,
} from "@material-ui/core";

import ImagesList from "../Images/List";
import { GROUP, IMAGES_BY_POST_ID } from "../../apollo/client/queries";
import UserAvatar from "../Users/Avatar";
import ItemMenu from "../Shared/ItemMenu";
import styles from "../../styles/Shared/Shared.module.scss";
import GroupItemAvatar from "../Groups/ItemAvatar";
import { Common, Roles } from "../../constants";
import {
  useCurrentUser,
  useHasPermissionGlobally,
  useUserById,
} from "../../hooks";
import { noCache, timeAgo } from "../../utils/clientIndex";
import CardFooter from "./CardFooter";

const useStyles = makeStyles({
  title: {
    marginLeft: "-5px",
  },
});

interface Props {
  post: Post;
  deletePost: (id: string) => void;
}

const Post = ({ post, deletePost }: Props) => {
  const { id, userId, groupId, postGroupId, body, createdAt } = post;
  const currentUser = useCurrentUser();
  const [canManagePosts] = useHasPermissionGlobally(
    Roles.Permissions.ManagePosts
  );
  const user = useUserById(userId);
  const [group, setGroup] = useState<Group>();
  const [images, setImages] = useState<Image[]>([]);
  const [menuAnchorEl, setMenuAnchorEl] = useState<null | HTMLElement>(null);
  const [getGroupRes, groupRes] = useLazyQuery(GROUP);
  const imagesRes = useQuery(IMAGES_BY_POST_ID, {
    variables: { postId: id },
    ...noCache,
  });
  const classes = useStyles();
  const router = useRouter();

  useEffect(() => {
    if (imagesRes.data) setImages(imagesRes.data.imagesByPostId);
  }, [imagesRes.data]);

  useEffect(() => {
    if (groupRes.data) setGroup(groupRes.data.group);
  }, [groupRes.data]);

  useEffect(() => {
    if (groupId || postGroupId) {
      getGroupRes({
        variables: { id: groupId ? groupId : postGroupId },
      });
    }
  }, [groupId, postGroupId]);

  const ownPost = (): boolean => {
    if (currentUser && user && currentUser.id === user.id) return true;
    return false;
  };

  const onGroupPage = (): boolean => {
    return router.asPath.includes(Common.ResourcePaths.Group);
  };

  return (
    <div key={id}>
      <Card>
        <CardHeader
          avatar={
            group && !onGroupPage()
              ? user && (
                  <GroupItemAvatar user={user} group={group} post={post} />
                )
              : user && <UserAvatar user={user} />
          }
          title={
            (!group || onGroupPage()) && (
              <>
                <Link href={`/users/${user?.name}`}>
                  <a>{user?.name}</a>
                </Link>
                <Link href={`${Common.ResourcePaths.Post}${id}`}>
                  <a className={styles.timeAgo}>{timeAgo(createdAt)}</a>
                </Link>
              </>
            )
          }
          action={
            <ItemMenu
              itemId={id}
              itemType={Common.ModelNames.Post}
              anchorEl={menuAnchorEl}
              setAnchorEl={setMenuAnchorEl}
              deleteItem={deletePost}
              ownItem={ownPost}
              hasPermission={canManagePosts}
            />
          }
          classes={{ title: classes.title }}
        />

        {body && (
          <CardContent>
            <Typography
              style={{
                marginTop: "-12px",
              }}
            >
              {body}
            </Typography>
          </CardContent>
        )}

        <CardActionArea>
          <CardMedia>
            <ImagesList images={images} />
          </CardMedia>
        </CardActionArea>

        {!!currentUser && <CardFooter postId={id} />}
      </Card>
    </div>
  );
};

export default Post;
