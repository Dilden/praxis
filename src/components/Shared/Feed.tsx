import { CircularProgress } from "@material-ui/core";
import Motion from "../Motions/Motion";
import Post from "../Posts/Post";

interface Props {
  feed: FeedItem[];
  deleteMotion: (id: string) => void;
  deletePost: (id: string) => void;
}

const List = ({ feed, deleteMotion, deletePost }: Props) => {
  if (feed)
    return (
      <div style={{ marginBottom: "200px" }}>
        {feed
          .sort((a, b) => parseInt(b.createdAt) - parseInt(a.createdAt))
          .map((item) => {
            return item.__typename === "Motion" ? (
              <Motion
                motion={item as Motion}
                deleteMotion={deleteMotion}
                key={`motion-${item.id}`}
              />
            ) : (
              <Post
                post={item as Post}
                deletePost={deletePost}
                key={`post-${item.id}`}
              />
            );
          })}
      </div>
    );

  return <CircularProgress style={{ color: "white" }} />;
};

export default List;
