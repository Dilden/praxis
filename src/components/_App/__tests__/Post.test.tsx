import { render, RenderResult, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { mocked } from "ts-jest/utils";
import { Common } from "../../../constants";
import Post from "../../Posts/Post";
import {
  useUserById,
  useHasPermissionGlobally,
  useWindowSize,
} from "../../../hooks";
import { mockNextUseRouter } from "../../Shared/__mocks__";
import deletePostHandler from "../../../pages/posts/[id]";

jest.mock("../../../hooks", () => ({
  useCurrentUser: jest.fn(),
  useUserById: jest.fn(),
  useHasPermissionGlobally: jest.fn(),
  useWindowSize: jest.fn(),
}));

jest.mock("../../../utils/clientIndex", () => ({
  redeemedInviteToken: jest.fn(),
}));

jest.mock("../../../utils/time", () => ({
  timeAgo: jest.fn(),
}));

const mockUserName = "mockUserName";
const mockUser: User = {
  id: "1",
  name: mockUserName,
  email: "test@email.com",
  createdAt: new Date().toLocaleDateString(),
  exp: 1,
};

const mockPost: Post = {
  id: "1",
  userId: "1",
  groupId: "",
  postGroupId: "",
  body: "This is a test post",
  createdAt: new Date().toLocaleTimeString(),
  __typename: Common.TypeNames.Post,
};

const renderPost = async (): Promise<RenderResult> => {
  mockNextUseRouter("/posts/" + mockPost.id);
  return render(
    <MockedProvider>
        <Post post={mockPost} deletePost={deletePostHandler} />
    </MockedProvider>
  );
};

describe("Post", () => {
  beforeAll(() => {
    mocked(useWindowSize).mockReturnValueOnce([]);
  });

  it("should display the user's name and post body", async () => {
    mocked(useHasPermissionGlobally).mockReturnValue([true]);
    mocked(useUserById).mockReturnValue(mockUser);

    const { getByText, container } = await renderPost();
    const userName = getByText(mockUserName);
    const postBody = getByText(mockPost.body);

    await waitFor(() => expect(userName).toBeInTheDocument());
    await waitFor(() => expect(postBody).toBeInTheDocument());

    expect(container).toMatchSnapshot();
  });
});
