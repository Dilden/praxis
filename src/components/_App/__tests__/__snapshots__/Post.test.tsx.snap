// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post should display the user's name, post body, and time 1`] = `
<div>
  <div>
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar"
        >
          <a
            href="/users/mockUserName"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular"
              style="color: black; background-color: white;"
            >
              <img
                alt="Data could not render."
                class="MuiAvatar-img"
                src="http://localhost:3000undefined"
              />
            </div>
          </a>
        </div>
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title makeStyles-title-1 MuiTypography-body2 MuiTypography-displayBlock"
          >
            <a
              href="/users/mockUserName"
            >
              mockUserName
            </a>
            <a
              class="timeAgo"
              href="/posts/1"
            >
               Â· now
            </a>
          </span>
        </div>
        <div
          class="MuiCardHeader-action"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1"
          style="margin-top: -12px;"
        >
          This is a test post
        </p>
      </div>
      <div
        class="MuiCardMedia-root"
      />
    </div>
  </div>
</div>
`;
